### ISSUE TITLE
[新功能] 自HomePod 1.0以来pytorch的更新

### ISSUE THREAD

**Issue by gemfield** [DeepVAC/MLab#7]
- fd02fc5d715a7647631c5806db736794edc2a52f: Port put_ and take from TH to ATen
- 4170a6cc24c6867ca6cd48f5581e98a4be89e593: Migrate `mode` from TH to ATen
- 6866c033d5aa134a83bc1cb84e3e084a7329167f: [JIT] Add recursive scripting for class type module attributes

**Comment by gemfield** [comment]
- 3551bd31bef401a08bf407cfe466814f1fbe250c: [PyTorch] Lite interpreter with a backend delegate
- a0d977610421823619ca376a3f51457bf4c0be82:  [JIT] Include conv3d in the conv-add-relu fusion
- NNAPI 系列
- 4757d4c077cb4d47793aae2fb89637bf8d3c0213: Don't allocate result Tensors in out xxxxxxx
- 34b46359e3cdca01f460fe8f4ce894868a2b4191: fix IValue Fix forwarding/move bug

**Comment by gemfield** [comment]
- 10abbb812a47bfabba4905b50dde02a511191522: Support tensor subclasses in Torchscript

**Comment by gemfield** [comment]
- 2564c0c88974bdfc13321c3d4fdffab454302c81: avoid CPU std::copysign segfault when compiling on arm64
- 305abde976d232494c6b5252111ce15523511ffe: Fix nvcc warnings
- 42486963b23868d5b423165aa56de2e5547d841e: Integrate NNC conv2d with fuser
- d2784c233bfc57a1d836d961694bcc8ec4ed45e4: Partially migrate sort from THC to ATen, replace the thrust path with cub
- bbd2b1bd3ca3e84901afd336194024fca5c44b1c: [quant][graphmode][fx] Add shape to nontensor op list
- torch/elastic stufff
- NNC stuff
-

**Comment by gemfield** [comment]
- 84a7ab250bd1db6ff72d8b0e01f046e9955da65f: Optimize constructing tensors from external data
- c6d9ca0c2b4bb124a902db00393372c0fc12b190: [reland]Replace AutoNonVariableTypeMode with InferenceMode in static runtime.
- 24f9a446c9ddb9442e41f254d0884d417e7fd76a: Fix wrong detection of depthwise conv on neon
- aceceb3d5c4f8b0632c01b75d6c51c08609cb131: Reland #50999 (Added pow() on CPU for float16 & bfloat16)
- 1688a5d31a740fe502232295ec22e49d376e4f2d: Cleanup since FEATURE_TORCH_MOBILE is always true.
- 88c06d9dfce27f49bc7cf0c308ba1cb0dabfc5b9: Add cuda device synchronization support in JIT
- f8d331b33ba6ebea20aee62f02dafe3ac9b69ea5: PyTorch Execution Graph Observers
- 81f181567a85dda92ba2ea24e0448ddebb9ea666: Add `USE_MAGMA` build flag
- 75b6644a4cc7d9c6848588f1804f98e46cc64ce1: Add USE_NUMPY define only if PyTorch is compiled with Numpy
- 9d3d169d2d1507aec7cca3ab67abb45f0059cfa0: Implement hardswish/hardsigmoid on MKLDNN tensors
- f26a6cb372569b957318c7f2a44c92428cc9686b: [quantization] Fix deepcopy on quantized ConvNd

**Comment by gemfield** [comment]
- 2c9972facf      Tao Xu  Mon Apr 19 13:51:37 2021 -0700  [iOS GPU][Kernel] Implement transpose in Metal shaders (#54522)
- 94406f77f6      Jerry Zhang     Mon Apr 19 21:36:01 2021 -0700  [quant][graphmode][fx] Add support for keeping output quantized for list and dict (#56391)
- 1f0223d6bb      M.L. Croci      Thu Apr 22 07:42:11 2021 -0700  Fix bug in gaussian_nll_loss (#56469)
- c37095760d      Liang Luo       Fri Apr 23 14:11:09 2021 -0700  [torch distributed] Implementing all_gather_base (#56315)

**Comment by gemfield** [comment]
- f84f2063b4      Ivan Yashchuk   Mon Apr 26 09:47:13 2021 -0700  Port CUDA torch.geqrf to ATen (#56251)
- d221be6fb4      Tao Xu  Mon Apr 26 16:33:19 2021 -0700  [iOS GPU] Use thread buffer to store indices for transpose (#56706)
- 502c58ad84      Vasiliy Kuznetsov       Mon Apr 26 16:58:35 2021 -0700  ns for fx:xxxxxxxxxx
- e909ad2dc4      Ansha Yu        Mon Apr 26 17:25:26 2021 -0700  [static runtime] binding for aten::argmin_out (#56638)

**Comment by gemfield** [comment]
- dc8a8cea79      Pritam Damania  Mon Apr 26 23:06:55 2021 -0700  Move caffe2 signal_handler to c10. (#56717)
- ab1457ad14      Edward Yang     Tue Apr 27 07:33:43 2021 -0700  Remove C++17 only optional include (#56782)
- 5b3c0ae563      Shen Li Tue Apr 27 12:27:44 2021 -0700  Use a FutureFactoryRegistry to allow libtorch_cpu files to create CUDAFuture (#56984)
- 522dca4ab0      Eddie Yan       Tue Apr 27 14:47:05 2021 -0700  Port `topk` from THC to ATen, migrate most of sort as well (#55392)
- fda8561944      Heitor Schueroff        Thu Apr 29 08:11:10 2021 -0700  Adding vector_norm to the C++ API (#57055)
- 9486fc3229      Chen Lai        Thu Apr 29 10:06:38 2021 -0700  [PyTorch][Edge] share readArchiveAndTensors between mobile and jit (#57098)
- e31b67f550      Michael Suo     Thu Apr 29 22:28:08 2021 -0700  [torch/deploy] opt torch/csrc/depoy into autofromatting
- 731cc472c5      leslie-fang-intel       Fri Apr 30 08:45:08 2021 -0700  refactor autocast to be extensible for devices (#57104)
- e62cdae469      Edvard Ghazaryan        Fri Apr 30 10:48:38 2021 -0700  Static Runtime support for aten::matmul (#57291)
- f54aa85a6c      Ivan Yashchuk   Fri Apr 30 11:07:31 2021 -0700  Fix MAGMA qr for empty batched inputs (#56257)
- 3a777b6792      Scott Wolchok   Fri Apr 30 11:19:41 2021 -0700  [PyTorch] Optimize intrusive_ptr(TTarget*) ctor (pybind) (#57053)
- c1a442248b      Bin Bao Fri Apr 30 15:26:13 2021 -0700  [JIT] Enable conv-add-relu fusion as a part of frozen graph optimization (#56580)
- 293830bc19      Sameer Deshmukh Fri Apr 30 15:51:37 2021 -0700  Fix min() and max() for empty tensors (#52565)
- 3a5f85465b      Jongsoo Park    Fri Apr 30 16:34:48 2021 -0700  [pytorch] fewer cuda sync in unique by using cub instead of thrust (#57323)
- 47e9ec401a      Horace He       Fri Apr 30 17:33:39 2021 -0700  [nnc] ported some more ops + added vectors to argvalue (#56766)
- bd347012ec      Mustafa Bal     Fri Apr 30 17:50:40 2021 -0700  Added sm_75 support for CI Xenial CUDA 11.1 cuDNN 8 builds (#57320)
- 3c4d57c18b      Jiakai Liu      Fri Apr 30 19:06:08 2021 -0700  [pytorch][nnc] update external functions for mobile build (#56850)
- 0787d781c5      Richard Zou     Wed May 5 07:35:04 2021 -0700   Fix compatibility problem with LSTMs and torch.save (#57558)
- a948e279ac      Rohan Varma     Wed May 5 10:12:20 2021 -0700   [c10d] Profiler support for nccl p2p collectives (#56427)
- 28cd04ea64      Akshit Khurana  Wed May 5 13:51:59 2021 -0700   NNAPI: add flexible size support for conv2d (#57561)
- c4bb6a5781      Akshit Khurana  Wed May 5 13:51:59 2021 -0700   NNAPI: flex size support for upsample_nearest2d op (#57563)
- 56211524a7      Horace He       Wed May 5 17:08:16 2021 -0700   [NNC] ported over sum and softmax to new scheme (#56775)
- c27428b5e9      Horace He       Wed May 5 20:53:25 2021 -0700   [nnc] ported conv2d lowering over (#56875)
- 86b061c80e      Horace He       Wed May 5 20:53:25 2021 -0700   [FX] Changes in order to move python key out of tree (#57427)
- ba78bf1363      Yi Huang (Symphony)     Wed May 5 21:40:07 2021 -0700   [standaloneRunner] fix another GIL mutithreading issue exposed by torch::jit::toIValue() (#57688)
- 7870450706      Scott Wolchok   Thu May 6 00:12:18 2021 -0700   [PyTorch] Use c10::ThreadLocal instead thread_local in record_function.cpp for specific __GLIBCXX__ on Android (#57689)
- 95fbc158d4      Raghavan Raman  Thu May 6 01:05:49 2021 -0700   [NNC] Add a method to compute conv without bias (#57512)
- 1f178de800      Raghavan Raman  Thu May 6 01:05:49 2021 -0700   [NNC] Add support for computing conv with dynamic shapes (#57514)
- d82333e92a      Jiakai Liu      Thu May 6 03:22:59 2021 -0700   [pytorch][nnc] protocol classes to persist the context for compiled functions (#56851);
- b4a098f1fb      Jiakai Liu      Thu May 6 03:22:59 2021 -0700   [pytorch][nnc] mobile nnc backend skeleton (#56852);
- 59d794b2c3      Ivan Yashchuk   Thu May 6 04:43:10 2021 -0700   Port CPU torch.ormqr to ATen (#57315);

**Comment by gemfield** [comment]
- aedcff7275      Rong Rong (AI Infra)    Thu May 6 15:59:50 2021 -0700   fix codegen for lite_interpreter (#57761)
- 241c2f4496      Elias Ellison   Thu May 6 16:03:30 2021 -0700   Add Gelu To NNC (#57753)
- 8b38458011      Zhengxu Chen    Thu May 6 16:42:41 2021 -0700   [jit] Break interpreter.cpp into smaller files. (#56546)
- e5e095cbe4      Can Balioglu    Thu May 6 19:50:38 2021 -0700   [torch/elastic] Rename etcd-/c10d-experimental to etcd-v2 and c10d (#57764)

**Comment by gemfield** [comment]
- 7179e7ea7b      Nikita Shulga   Tue May 25 15:08:56 2021 -0700  [CMake] Prefer third_party/pybind11 by default (#58951)
- 1bd22e28b3      Xiang Gao       Wed May 26 16:47:50 2021 -0700  BFloat16 support for torch.sort (#58196)
- 95b1bc1009      Peter Bell      Thu May 27 10:05:40 2021 -0700  Migrate nonzero from TH to ATen (CPU) (#58811)
- 4e543d017f      Eddie Yan       Thu May 27 18:34:37 2021 -0700  Move remaining \*Sort\* in `THC` to `ATen` (#58953)
- 9c83e4160d      David Reiss     Thu May 27 20:46:51 2021 -0700  Use some c10::ThreadLocal to avoid crashes on old Android toolchains (#59017)

**Comment by gemfield** [comment]
- 67b8e6410d      Tao Xu  Tue Jun 8 14:40:57 2021 -0700   [OSS] Add podspec for libtorch-lite (#59638)
- 9d1d799034      Hariom Narang   Mon Jun 21 16:09:35 2021 -0700  Added API to change logging levels for JIT (#58821)
- 729f7cd52f      Saketh Are      Tue Jun 22 10:04:51 2021 -0700  Implement histogram operator on CPU (#58780)
- 567e6d3a87      Winston Smith   Tue Jun 22 10:25:27 2021 -0700  Remove Caffe2 thread-pool leak warning (#60318)
- 0d2a936176      Ilqar Ramazanli Tue Jun 22 10:37:31 2021 -0700  To add Rectified Adam Algorithm to Optimizers (#58968)
- 31a884987d      Peter Bell      Tue Jun 22 10:53:55 2021 -0700  Remove some TH includes from ATen (#60323)
- 99ca2c5b4b      Thomas J. Fan   Tue Jun 22 12:02:12 2021 -0700  Migrates nll_loss_backward from TH to Aten (CUDA) (#60299)
- e8690dacb2      Ilqar Ramazanli Wed Jun 23 08:20:06 2021 -0700  To add Nesterov Adam Algorithm to Optimizers (#59009)
- 1120a1b92e      Supriya Rao     Wed Jun 23 15:53:55 2021 -0700  [quant][fx][fix] QAT with object_type in qconfig (#60555)
- a016150163      Luca Wehrstedt  Thu Jun 24 12:37:29 2021 -0700  Move torch/lib/c10d to torch/csrc/distributed/c10d (#60543)
- d586248544      Peter Bell      Thu Jun 24 19:39:36 2021 -0700  Migrate THStorage_resizeBytes to ATen (CPU) (#60324)
- 4a7d281119      Peter Bell      Thu Jun 24 19:39:36 2021 -0700  Migrate THAllocator to ATen (#60325)
- 42c8439b6e      Peter Bell      Thu Jun 24 19:39:36 2021 -0700  TH: Clean up dead code (#60655)
- 9b94aa5356      Angela Yi       Mon Jun 28 15:20:48 2021 -0700  [quant][fx][fix] Fused modules with object_type in qconfig (#60779)
- 287c0ab170      Malay Bag       Tue Jun 29 23:04:07 2021 -0700  [FX] Add requires_grad to TensorMetadata (#60972)

**Comment by gemfield** [comment]
- 46595a9623      Hao Lu  Thu Jul 1 00:00:36 2021 -0700   [Static Runtime] Add gflag to disable nnc and caffe2 math library (#61090)
